crates/vm/src/lib.rs

```diff
@@ -0,0 +1,443 @@
+pub struct VM<'w> {
+    output: &'w mut dyn io::Write,
+    stack: Vec<Value>,
+    heap: Heap,
+    frames: Vec<Frame>,
+}
+
+struct Frame {
+    func: FuncKey,
+    ip: usize,
+    stack_offset: usize,
+}
+
+impl<'w> VM<'w> {
+    pub fn new(output: &'w mut dyn io::Write) -> Self {
+        Self {
+            output,
+            stack: Default::default(),
+            heap: Default::default(),
+            frames: Default::default(),
+        }
+    }
+
+    pub fn stack_len(&self) -> usize {
+        self.stack.len()
+    }
+
+    fn ip(&self) -> usize {
+        self.frames.last().unwrap().ip
+    }
+
+    fn ip_mut(&mut self) -> &mut usize {
+        &mut self.frames.last_mut().unwrap().ip
+    }
+
+    fn stack_offset(&self) -> usize {
+        self.frames.last().unwrap().stack_offset
+    }
+
+    fn get_func<'b, 'a, 's>(&self, code: &'b bytecode::Code<'a, 's>) -> &'b bytecode::Func<'a, 's> {
+        let key = self.frames.last().unwrap().func;
+        &code.fns[key]
+    }
+
+    pub fn run(&mut self, code: &bytecode::Code, main_key: bytecode::FuncKey) -> Result<i64> {
+        // All of this assumes a fresh VM, which is probably bad design
+
+        self.frames.push(Frame {
+            func: main_key,
+            ip: 0,
+            stack_offset: 0,
+        });
+
+        // if DEBUG {
+        //     for (idx, el) in main.code.iter().enumerate() {
+        //         eprintln!("{} {:?}", idx, el);
+        //     }
+        //     eprintln!();
+        // }
+
+        loop {
+            let ip = self.ip();
+            let instr = &self.get_func(code).code[ip];
+
+            *self.ip_mut() += 1;
+            if DEBUG {
+                eprintln!(
+                    "{:?}",
+                    value::StackDbg {
+                        stack: &self.stack,
+                        heap: &self.heap
+                    }
+                );
+                eprintln!("{:?}", instr);
+                eprintln!();
+            }
+
+            match instr {
+                Instr::LoadLit(l) => {
+                    let val = match *l {
+                        Literal::String(s) => (self.add_to_heap(BigValue::String(s.to_owned()))),
+                        Literal::Float(f) => Value::Float(f),
+                        Literal::Integer(i) => Value::Int(i),
+                        Literal::Bool(b) => Value::Bool(b),
+                        Literal::Null => Value::Null,
+                    };
+                    self.push(val);
+                }
+                Instr::BinOp(o) => {
+                    let rhs = self.pop();
+                    let lhs = self.pop();
+
+                    // In the fast path, we dont encounter a type error, and
+                    // therefor dont need to access spans. We should only do
+                    // this load once we know its type error, for better cache.
+                    // TODO: do this.
+                    let ast = self.get_func(code).spans[ip].as_expr().unwrap();
```

Comment:
`self.get_func(code).spans[ip]` should be a a function, though you might run into borrowcheck problems

Reply:
Yep, full of borrowck problems